Step3 - 로또 (자동) 구현

devsigner9920 opened the pull request:
안녕하세요, 

이번주 야근이 잦아 늦게 PR 드립니다 ㅠ

그렇다고 쉽게 쉽게 생각하고 프로그램 짜지 않고 열심히 하겠습니다!

피드백 부탁드립니다. 감사합니다 :)

devsigner9920 committed:
docs(README): 로또(자동) 프로그램 기능 목록 정의

변경 내용:
- 로또(자동) 프로그램 구현을 위해 기능 목록 및 기능 정의를 작성하였습니다.

devsigner9920 committed:
feat(Console): 콘솔 유틸 추가

추가 내용:
- 프리 코스에서 사용한 Console 유틸을 추가하였습니다.

devsigner9920 committed:
feat(LottoGenerator): 로또 번호 생성 객체 추가

추가 내용:
- 로또 번호를 생성하는 방법이 추후에 변경될 수 있으므로 interface 형태로 생성 후 구현체를 생성하였습니다.
- 구현체는 현재 요건인 자동 생성에 맞는 구현체를 작성하였습니다.

devsigner9920 committed:
feat(constants): 상수 클래스 추가

추가 내용:
- 에러 메시지 및 결과 출력 메시지를 추가하였습니다.

devsigner9920 committed:
feat(PurchaseAmount, LottoBuyer): 로또 구매자 객체 추가

추가 내용:
- 로또 구매자 객체 및 구매자가 지불한 금액에 대한 객체를 추가하였습니다.
- 로또 구매자는 구매 행위 및 당첨 결과와 본인이 구매한 로또와 매치하는 로직을 가지고 있습니다.

devsigner9920 committed:
feat(Lotto): 로또 객체 추가

추가 내용:
- 로또 및 로또 일급 컬렉션을 구현하였습니다.
- 로또 객체는 다른 로또와 몇개의 숫자가 맞는지 갯수를 구할 수 있습니다.
- 로또들 객체는 종합 당첨 결과를 반환해줄 수 있습니다.

devsigner9920 committed:
feat: 로또 게임 진행을 위한 제어와 서비스를 제공하는 클래스 구현

추가 내용:
- LottoApplication: 로또 프로그램 시작 클래스 생성하였습니다.
- LottoDrawer: 당첨결과를 뽑기위한 객체입니다.
- LottoController: 로또 게임을 진행하기 위한 설정, 게임 시작을 담당합니다.
- InputView: 로또 프로그램에서 받는 입력값을 처리합니다.

devsigner9920 committed:
feat(LottoRank): 로또 당첨 결과 정보를 가지고 있는 enum 객체 생성

devsigner9920 committed:
feat(MatchResult): 로또 당첨 결과를 가지고 있는 클래스 생성

추가 내용:
- 로또 당첨 결과 및 그 결과에 대한 당첨 횟수를 가진 클래스를 생성하였습니다.
- 일급 객체는 모든 당첨 결과를 종합하여 통계치 및 총 당첨 금액을 반환하는 로직을 구현하였습니다.

devsigner9920 committed:
feat: 로또 게임 결과 출력 및 실행 서비스 생성

5minho commented on src/main/java/lotto/generator/LottoGenerator.java:
```diff
@@ -0,0 +1,24 @@
+
+  Lotto generate();
+
+  default List<Integer> getLottoNumbers() {
+    List<Integer> lottoNumbers = new ArrayList<>();
+    for (int i = MIN_LOTTO_NUMBER; i <= MAX_LOTTO_NUMBER; i++) {
+      lottoNumbers.add(i);
+    }
+
+    return lottoNumbers;
```
이 메서드와 로또가격, 로또 최소 최대 숫자들 같은 변수들이 인터페이스에 정의되어야할 이유가 있나요?
로직이 안에 어떻게 구현되어 있는지(세부사항)은 인터페이스가 아닌 구현체가 알고있어야 하는 내용이라 생각 됩니다.
그리고 이렇게 되면 메서드가 public 으로 노출이 되는데 1 부터 45 까지 숫자 리스트를 만드는 것을 LottoGenerator 를 사용하는 클라이언트가 알 필요가 없는거 같아요.
구현체에 정의하는 것은 어떨까요?

devsigner9920 replied:
아하 처음 의도는 LottoGenerator가 로또를 생성하는 방법에 대해서만 여러 구현체를 가지겠지 라고 생각했습니다! ex) 수동 생성, 자동 생성 등등..

그래서 수동 생성이든 자동 생성이든 로또 번호를 생성하는 최소 최대 값과 번호 목록을 LottoGenerator가 가지는게 나을거 같다고 생각했습니다ㅠ

private 하지 않고 노출되어 있어서 안전하지 않은 것과 인터페이스는 인터페이스 그 자체로서 역할만 한다는 것을 간과한 것 같습니다!

좋은 의견 감사합니다!

5minho commented on src/main/java/lotto/generator/AutoLottoGenerator.java:
```diff
@@ -0,0 +1,25 @@
+    return lottoNumbers.subList(ZERO, LOTTO_NUMBER_RANGE);
+  }
+
+  private List<Integer> shuffle(List<Integer> lottoNumbers) {
+    Collections.shuffle(lottoNumbers);
+    return lottoNumbers;
```
해당 메서드가 필요한 이유는 무엇인가요?
peakLottoNumbers 처럼 의도를 더 명확히 하려는 것도 아닌거 같고,
내부에서 shuffle 외에 딱히 추가되는 로직도 없고, 
반환값도 매개변수도 그대로 리턴을 해서요.

devsigner9920 replied:
제가 생각했을 땐 따로 빼두는게 나을 것 같아 했는데 너무 쓸데없이 나눈 것 같습니다..! 수정하겠습니다!

5minho commented on src/main/java/lotto/model/LottoDrawer.java:
```diff
@@ -0,0 +1,23 @@
+
+import static lotto.view.OutputView.*;
+
+public class LottoDrawer {
```
해당 클래스는 model 패키지 내에 있으면서 View 에 의존성을 같네요.
MVC 패턴을 사용해서 구현해 보신거 같은데 MVC 패턴에서 Model 은 View 를 몰라야 합니다.

devsigner9920 replied:
아하 넵 알겠습니다 감사합니다!!

5minho commented on src/main/java/lotto/model/LottoDrawer.java:
```diff
@@ -0,0 +1,23 @@
+    this.lotto = lotto;
+  }
+
+  public static LottoDrawer draw(LottoGenerator lottoGenerator) {
```
이 메서드는 너무 여러가지 일을 하고 있다고 생각해요.
draw 내부에서 Lotto 를 생성하고, 그걸 프린트 하고, LottoDrawer 라는 객체를 반환까지 합니다.
그리고 제 생각에는 위의 로직은 model 클래스가 할 일이 아닌 controller 의 역할이라고 생각합니다.

devsigner9920 replied:
넵 그럼 컨트롤러에서 해당 로직을 처리하도록 해보겠습니다!

5minho commented on src/main/java/lotto/model/MatchResults.java:
```diff
@@ -0,0 +1,35 @@
+    this.matchResults = matchResults;
+  }
+
+  public void statistics(PurchaseAmount purchaseAmount) {
+    OutputView.printStatisticsGuideMessage();
+    OutputView.printMatches(matchResults);
+    OutputView.printYield(calculateYield(purchaseAmount));
+  }
```
해당 클래스도 model 내에 view 의존성을 없애보면 좋을거 같아요

devsigner9920 replied:
앗 넵 알겠습니다!

5minho commented on src/main/java/lotto/util/Console.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.Objects;
+import java.util.Scanner;
+
+/**
+ * @author Jung YoonSung <ybook2012@gmail.com>, Kim HyeonSik <hsik0225@gmail.com>, Kim JuWon <kjw11077naver@gmail.com>
+ * @version 1.0
+ * @since 1.0
+ */
+public class Console {
```
이 클래스는 존재의 이유가 무엇인가요?
원준님이 작성하신건가요?

devsigner9920 replied:
@5minho 프리코스 당시에 활용했던 콘솔 유틸리틴데 실제 프로그램 동작을 하기 위해 입력값을 받는 유틸입니다!

5minho replied:
원준님이 작성하신건가요? author 주석을 보면 원준님께서 작성하신 코드가 아닌거 같아서요.

devsigner9920 replied:
아뇨! 제가 작성한건 아닙니다 혹시 프리코스에서 제공한 코드 그대로 사용하면 안될까요? 안된다면 새로 작성하겠습니다!

5minho replied:
아 NextStep 에서 제공한 코드이군요! 그 부분은 제가 몰랐네요 😂😂
이번 미션에서 크게 영향을 주는 코드도 아니고, NextStep 에서 제공한 코드이니 새로 작성해주실 필요는 없습니다~

devsigner9920 replied:
앗! 그래도 저도 좀 찜찜해서 새로 작성했습니다!! ㅎㅎㅎ 매번 리뷰 꼼꼼히 해주셔서 감사합니다! 해당 부분 설명없이 쓴 제 잘못입니다 ㅠ 새로 작성하고 다른 피드백도 반영하여 두었습니다! 감사합니다 ㅎ

5minho commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.List;
+
+public class Lotto {
```
로또의 6개의 숫자를 가지고 있는 일급컬렉션 같은데요
더 명확하게 LottoNumbers 나 LottoNumber 라고 해보는건 어떨까요?

devsigner9920 replied:
아... 로또 숫자를 가지고 있으니.. 아하 알겠습니다 네이밍 아이디어 감사합니다!

5minho commented on src/main/java/lotto/constants/OutputMessage.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.constants;
+
+public class OutputMessage {
+  public static final String INPUT_PURCHASE_AMOUNT_GUIDE_MESSAGE = "구입금액을 입력해 주세요.";
+  public static final String INPUT_BUY_LOTTO_COUNTS_MESSAGE = "%d개를 구매했습니다.";
+  public static final String LAST_WEEK_DRAWN_LOTTO = "지난 주 당첨 번호를 입력해 주세요.";
+  public static final String STATISTICS_GUIDE_MESSAGE = "당첨 통계";
+  public static final String PERFORATION = "-------";
+  public static final String MATCH_RESULT_MESSAGE = "%d개 일치 (%d원)- %d개";
+  public static final String YIELD_MESSAGE = "총 수익률은 %.2f입니다.";
+}
```
해당 OutputMessasge 가 모두 public 으로 공개되어 있네요.
View 내부에서만 쓰는 문자열들일것 같은데 이렇게 공개되는 것보다 OutputView 내에 숨기는게 더 좋지 않을까요?

devsigner9920 replied:
아하.. 상수 메시지를 constants 패키지에 무조건 모아두는것이 좋다고 생각했는데 OutputView에서만 쓰인다면 OutputView 내에만 존재해도 되겠군요 감사합니다.

5minho commented on src/main/java/lotto/constants/LottoRank.java:
```diff
@@ -0,0 +1,33 @@
+    return Arrays.stream(LottoRank.values())
+      .filter(lottoRank -> lottoRank.rank == rank)
+      .findAny()
+      .orElse(null);
```
rank 에 일치하는 enum 이 없으면 예외인 상황 같은데 이 경우에는 null 을 리턴하는 것보다 예외를 던지는게 좋지 않을까요?

devsigner9920 replied:
넵 알겠습니다! 긴가민가 했는데 예외를 던지는 것이 더 나은거 같군요 감사합니다!!!

5minho commented on src/main/java/lotto/LottoController.java:
```diff
@@ -0,0 +1,23 @@
+
+import static lotto.view.InputView.*;
+
+public class LottoController {
```
지금 기능요구사항에 부합하지 않는거 같아요.
사용자가 로또번호를 입력해야 하는데 자동으로 생성해버리네요.

devsigner9920 replied:
@5minho 앗 제가 3단계 미션 요구사항 이해를 했을 땐 자동으로 로또 번호를 발급하는걸로 이해해서요! 혹시 아닐까요..?

5minho replied:
지난주 당첨 번호를 입력해주세요.
라고 메시지가 나오면 사용자가 당첨 번호를 입력해야 하는데, 자동 입력되어 넘어갑니다ㅜ

devsigner9920 replied:
@5minho 아하.. 제가 잘못 이해했네요 ㅠ 해당부분 입력받도록 수정하겠습니다!

5minho commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Lottos {
```
여러개의 6개의 로또 숫자들을 담는 일급컬렉션 같은데요.
제 생각에는 하나의 로또티켓에 여러 로또 숫자들이 들어가니 LottoTicket 이라고 해보는 것도 좋을거 같아요.

devsigner9920 replied:
넵 클래스 명명에 대해서 좀 더 생각하고 명명해야겠네요! 단순하게 복수형을 붙이는 것 보단 이런식으로 하겠습니다!! ㅎㅎ

5minho reviewed (changes requested):
안녕하세요 원준님.
미션하시느라 고생많으셨습니다.
피드백 몇 개 남겼는데 확인부탁드려요.

devsigner9920 committed:
refactor(naming): 리뷰어님의 피드백 의견 반영

변경 내용:
- 리뷰어님의 피드백 의견을 적극 반영하여 코드를 조금 더 관심사항을 분리 시키고 네이밍에 신경써보았습니다.
- 굳이 필요 없는 클래스들을 삭제하고 조금 더 메소드들을 알맞는 위치에 쓰는 것을 신경썼습니다.

devsigner9920 committed:
refactor(naming): 리뷰어님의 2차 피드백 의견 반영

변경 내용:
- 리뷰어님 의견을 반영하여 요구사항에 맞게 당첨 로또를 자동 생성하지않고 직접 입력받을 수 있도록 구성하였습니다.
- Console 을 InputUtils로 변경하여 작성하였습니다.

5minho commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,96 @@
+public class InputView {
+  private static final String SPLIT_DELIMITER = ",";
+  private static Pattern numberPattern = Pattern.compile("\\d+");
+
```
InputUtils 를 사용하는 것 대신에 InputView 에
```
private static final Scanner SCANNER = new Scanner(System.in);
```
이렇게 선언해서 사용하셔도 될거 같아요

devsigner9920 replied:
아하 넵 알겠습니다!

5minho commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,96 @@
+  private static PurchaseAmount inputPurchaseAmountAndVerify() {
+    String input = InputUtils.readLine();
+    try {
+      checkNullOrEmpty(input);
+      checkStringIsNumber(input);
+      return new PurchaseAmount(parseInt(input));
```
PurchaseAmount 에 정적 팩터리 메서드를 만들어서 input 에 대한 유효성 처리를 해당 객체에게 위임하는건 어떨까요?



```
try {
      return PurchaseAmount.valueOf(input);
    } catch (Exception e) {
      System.out.println(e.getMessage());
      return inputPurchaseAmountAndVerify();
    }
```

devsigner9920 replied:
앗.. 제가 아이디어가 부족했네요 더 좋은것 같습니다! 감사합니다 ㅎㅎ 이번에 확실히 습득하고 가겠습니다

5minho commented on src/main/java/lotto/LottoController.java:
```diff
@@ -0,0 +1,40 @@
+    PurchaseAmount purchaseAmount = inputPurchaseAmount();
+    printBuyCount(purchaseAmount);
+
+    LottoBuyer lottoBuyer = LottoBuyer.buy(purchaseAmount, lottoGenerator);
```
LottoBuyer.buy() 의 반환형이 LottoBuyer 인게 이해가 잘안가네요 😅 LottoTicket 을 반환하도록 해보면 어떨까요?

5minho commented on src/main/java/lotto/model/LottoTicket.java:
```diff
@@ -0,0 +1,49 @@
+    return new MatchResults(matchResults);
+  }
+
+  private void countingWinningResultsByOrder(LottoNumbers winningLottoNumbers, MatchResult matchResult) {
+    for (LottoNumbers lottoNumbers : lottoTicket) {
+      countMatching(winningLottoNumbers, matchResult, lottoNumbers);
+    }
+  }
+
+  private void countMatching(LottoNumbers winningLottoNumbers, MatchResult matchResult, LottoNumbers lottoNumbers) {
+    if (matchResult.isWin(lottoNumbers.countMatchNumber(winningLottoNumbers))) {
+      matchResult.addCount();
+    }
+  }
```
메서드 내에서 인자로 받은 객체의 데이터를 바꾸는 것은 호출하는 입장에서 별로 좋지 않습니다.
명확하게 countXXXX 의 결과가 숫자 또는 MatchCount 가 되도록 리팩토링 해보면 좋을거 같습니다.

devsigner9920 replied:
아하... 그럼 그렇게 바꿔보겠습니다!

5minho commented on src/main/java/lotto/constants/LottoRank.java:
```diff
@@ -0,0 +1,33 @@
+  THREE_MATCHES(3, 5000),
+  FOUR_MATCHES(4, 50000),
+  FIVE_MATCHES(5, 1500000),
+  SIX_MATCHES(6, 2000000000);
```
```
SIX_MATCHES(6, 2_000_000_000);
```
 이렇게 언더스코어도 이용하면 금액 파악이 더 수월할거 같아요

devsigner9920 replied:
아하 넵! 알겠습니다!

5minho commented on src/main/java/lotto/model/MatchResults.java:
```diff
@@ -0,0 +1,27 @@
+  public long getTotalWinningAmount() {
+    long sum = 0;
+    for (MatchResult matchResult : matchResults) {
+      sum += ((long) matchResult.getMatchCount() * matchResult.getLottoRank().getMoney());
```
MatchResult 내에서 getter 로 변수를 꺼내와서 계산하는 것보다.
MatchResult 에게 메시지를 보내 계산하게 하면 어떨까요?
```
sum += matchResult.calculateWinningAmount();
```

devsigner9920 replied:
아하 넵 알겠습니다! 그러네요.. 매치 리절트 내에서 다 할 수 있었네요..

5minho commented on src/main/java/lotto/model/MatchResults.java:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  public List<MatchResult> getMatchResults() {
+    return matchResults;
```
해당 메서드를 호출한 클라이언트에서 컬렉션을 변경하면 어떤일이 벌어질까요?
값을 복사해서 리턴하면 사이드 이펙트를 줄일 수 있을거 같아요!

devsigner9920 replied:
클라이언트에서 변조가 생기지 않도록 리턴하라는 말씀인것 같아 그렇게 적용해보겠습니다 감사합니다!

5minho reviewed (approved):
안녕하세요 원준님!
피드백 반영하시느라 고생하셨습니다!
추가 코멘트 남겨봤는데요~ 다음 단계 진행하시면서 확인해주시면 감사하겠습니다!
